Undo redo feature

leroytan opened the pull request:
Close #56

leroytan committed:
Merge branch 'master' of https://github.com/AY2425S1-CS2103T-G19-4/tp

leroytan committed:
Merge branch 'master' of https://github.com/AY2425S1-CS2103T-G19-4/tp

leroytan committed:
Add undo and redo functionality to address book

* Implement new AddressBook with
version to track the history of changes.
* Add "commit", "undo", and "redo" methods to AddressBook with version.
* Update ModelManager to use AddressBook with version for
undo/redo operations.
* Create UndoCommand and RedoCommand classes.
* Add tests for UndoCommand in UndoCommandTest.
* Add tests for RedoCommand in RedoCommandTest.

leroytan committed:
Add user guide documentation for undo and redo command

Add descriptions and examples for `:undo` and `:redo` commands.
Include concrete examples to illustrate the usage of undo and redo
functionalities.

leroytan committed:
Merge branch 'master' into undo-redo-feature

rachelyeohm commented on src/test/java/seedu/address/logic/commands/UndoCommandTest.java:
It would be good to have a test on executing undo more than 10 times / the capacity of the list of address books, and check for failure.

rachelyeohm commented on src/test/java/seedu/address/logic/commands/AddCommandTest.java:
Since there are now a lot more tests that give the assertion error "This method should not be called.", to reduce repetition, it could be good to condense it into a variable and call that variable each time.

rachelyeohm commented on docs/UserGuide.md:
```diff
@@ -147,8 +147,31 @@ Examples:
+Format: `:undo`
+
+Examples:
+* Adding a new contact with `:add -n John Doe -p 98765432 -e johnd@example.com -l John street, block 123, #01-01`, 
```
Just a minor grammar issue - "If we add a new contact" might be better.

anselmlong reviewed (commented):
LGTM, mostly just naming changes to versioned address book.

leroytan committed:
Add capacity test for undo and redo

Check for failure of undo and redo test when undo or redo
beyond the capacity.
Update user guide to reduce confusion for undo and redo.

leroytan committed:
Merge branch 'undo-redo-feature' of https://github.com/leroytan/tp into undo-redo-feature

leroytan committed:
Merge branch 'master' of https://github.com/AY2425S1-CS2103T-G19-4/tp into undo-redo-feature

leroytan committed:
Merge branch 'master' of https://github.com/AY2425S1-CS2103T-G19-4/tp

leroytan committed:
Merge branch 'master' into undo-redo-feature

leroytan committed:
Fix git conflict

Change the method name to resolve git conflict.
Model uses the new method name.
